Update docs for eager rent collection (#10348)
solana-labs/solana: docs/src/apps/rent.md

# Storage Rent for Accounts

Keeping accounts alive incurs storage cost called _rent_ because the cluster must actively maintain the data to process any future transactions on it. This is not like Bitcoin or Ethereum, where account storages don't incur any costs.

> The rent is debited from account's balance by the runtime upon the first access in the current epoch by transactions or once per an epoch if there is no transaction. The rent is up-front for the next epoch. If an account can't pay rent for the next epoch, it's purged immediately. The fee is currently a fixed rate, measured in bytes-times-epochs. The fee may change in the future.

Reviewer: "...if there is no transaction" -> "...if there are no transactions"
"...rent is up-front for the next epoch..." -> "..rent is paid up front for the next epoch..."
To be clear, is rent checked at the beginning of each epoch?  So if an account can't pay rent it will be purged immediately at the start of a new epoch?

Author: Thanks for correction!

> So if an account can't pay rent it will be purged immediately at the start of a new epoch?

_In principle_, yes. Sadly, current implementation has some edge cases: https://github.com/solana-labs/solana/issues/10426

> Also, there is a way to avoid paying rent with enough account balance called _rent-exempt_.

Reviewer: "Accounts can be _exempt_ from paying rent by maintaining a minimum balance, this _rent-exemption_ is described below."
Just thought to reframe away from 'avoiding' paying rent.

Note: The rent rate can change in the future. And this applies to the testnet and mainnet-beta.

As of writing, the fixed rent fee is 19.055441478439427 lamports (bytes-epochs). And an epoch is roughly 2 days.

Firstly, the rent calculation considers the size of account including the metadata including its address, owner, lamports, etc. Thus the rent fee starts from 128 bytes as the minimum to be rented even if an account has no data.

For example, if an no-data account is created with the initial transfer of 10,000 lamports. The rent is immediately debited from it on creation, resulting in the balance of 7,561 lamports.

You can calculate like this:

```
7,561 = 10,000 (= transfered lamports) - 2,439 (= this account's rent fee for a epoch)
2,439 = 19.055441478439427 (= rent rate) * 128 bytes (= minimum account size) * 1 (= epoch)
```

And the account balance will be reduced to 5,122 lamports at the next epoch even if there is no activity:

```
5,122 = 7,561 (= current balance) - 2,439 (= this account's rent fee for a epoch)
```

This also indicates an account will be immediately removed after creation if the transferred lamports is less than or equal to 2,439.

## Rent exemption

> Alternatively, an account can be exempt from rent collection entirely by depositing certain amount of lamports. Such minimum amount is defined as the 2 years worth of rent fee.

Reviewer: Is this minimum amount checked every epoch? Can it be removed during the epoch and transferred just when rent is checked?

Author: Good question! Fortunately, there is no loophole like this you're concerned. :)

> Alternatively, an account can be made entirely exempt from rent collection by depositing at least 2 years-worth of rent. This is checked every time an account's balance is reduced and rent is immediately debited once the balance goes below the minimum amount.

Author: @aeyakovenko @t-nelson A bit dated discord discussion but this is still true... As the pr's text implicitly says, if the minimum amount threshold changes due to underlying rent fee change (heh, this is still not a thing), the once rent-exempt account may become not exempt suddenly.

Should I create an issue for this? I couldn't found one: https://discord.com/channels/428295358100013066/489504501749907468/671877112638210069

```
[8:56 AM]  trent: If its balance falls below rent-exempt min balance it will be charged again.  Today that's managed per-program
[8:56 AM]  trent: @anatoly :point_up:
[9:32 AM]  anatoly: @trent is rent-exempt min updated after creation?
[9:36 AM]  trent: @anatoly I don't see where lamports_per_byte_year is ever changed today, so no
[9:37 AM]  anatoly: @trent is that a global?
[9:37 AM]  anatoly: When do we check that it fell below?
[9:38 AM]  anatoly: as a developer, do I have any way to create a state that lasts forever regardless of rent cost changing in the future?
[9:38 AM]  trent: That's a member of the Rent struct, it's assigned a global by default though https://github.com/solana-labs/solana/blob/master/sdk/src/rent.rs#L35
[9:39 AM]  trent: I don't believe rent-exemption is hardened against the price of rent moving today, no?
[9:40 AM]  anatoly: That’s busted.
[9:41 AM]  anatoly: Something got lost in translation :slight_smile:
```

Reviewer 2: @ryoqun It's mentioned here, https://github.com/solana-labs/solana/issues/8145.  Might be worth changing the title to something more likely to come up in search

Author: Thanks for the tip!

> # Storage Rent for Accounts

Author: @CriesofCarrots Finally, I've added the content. Could you specifically reivew this file to check whether this meets being a straightforward, functional, user-driven introduction to rent with topic of ` developers and other users who just want to know when and how and how much rent will be assessed to their accounts.`. :) I had that in mind while writing this.

> Keeping accounts alive incurs storage cost called _rent_ because the cluster must actively maintain the data to process any future transactions on it. This is not like Bitcoin or Ethereum, where account storages don't incur any costs.

Reviewer 3: ```suggestion
Keeping accounts alive on Solana incurs a storage cost called _rent_ because the cluster must actively maintain the data to process any future transactions on it. This is different from Bitcoin and Ethereum, where storing accounts doesn't incur any costs.
```

> The rent is debited from an account's balance by the runtime upon the first access in the current epoch by transactions or once per an epoch if there are no transactions. The rent is paid upfront for the next epoch. If an account can't pay rent for the next epoch, it's purged immediately at the start of a new epoch. The fee is currently a fixed rate, measured in bytes-times-epochs. The fee may change in the future.

Author: @CriesofCarrots Well, my understanding is based on this existing document: https://github.com/solana-labs/solana/pull/10348/files#diff-80339caf6f22201fa3399a7761c91b88L7

Maybe, I need to do a fact check for this.

> The rent is debited from an account's balance by the runtime upon the first access in the current epoch by transactions or once per an epoch if there are no transactions. The rent is paid upfront for the next epoch. If an account can't pay rent for the next epoch, it's purged immediately at the start of a new epoch. The fee is currently a fixed rate, measured in bytes-times-epochs. The fee may change in the future.

Author: @CriesofCarrots I took another try to explain while avoiding the confusion: https://github.com/solana-labs/solana/pull/10348/files#r436269646

This is rather human perception problem of how we interpret the definition of rent duration. Technically, rent calculation is the same, just we're explaining it from different angles.

`cover the remainder of the current epoch` thinking is also acceptable, if we special-case the epoch where the account is created and deleted. But, I wanted to emphasis that rent costs are constant/consistent with the equally-sized durations. So, I want to introduce the notion of `next/upcoming epoch`, mildly influenced by the original specification: https://github.com/solana-labs/solana/pull/10348/files#diff-80339caf6f22201fa3399a7761c91b88L7

How does the new explanation look to you? I want to hear your opinion.

Accounts can be exempt from paying rent by maintaining a minimum balance, this rent-exemption is described below.

## Calculation of rent

Note: The rent rate can change in the future. And this applies to the testnet and mainnet-beta.

> As of writing, the fixed rent fee is 19.055441478439427 lamports per byte-epoch. And an epoch is roughly 2 days.

Reviewer 3: ```suggestion
As of writing, the fixed rent fee is 19.055441478439427 lamports per byte-epoch on the testnet and mainnet-beta clusters. An [epoch](../terminology.md#epoch) is roughly 2 days.
```

Reviewer 3: Users may wonder about this strange number (especially as lamports can't be split). How is this calculated?

Author: Ok, I'll document this as well. :)

Author: done!

> Firstly, the rent calculation considers the size of an account including the metadata including its address, owner, lamports, etc. Thus the rent fee starts from 128 bytes as the minimum to be rented even if an account has no data.

Reviewer 3: ```suggestion
The rent calculation includes account metadata (address, owner, lamports, etc) in the size of an account. Therefore the smallest an account can be for rent calculations is 128 bytes.
```

> For example, if a no-data account is created with the initial transfer of 10,000 lamports. The rent is immediately debited from it on creation, resulting in a balance of 7,561 lamports.

Reviewer 3: ```suggestion
For example, an account is created with the initial transfer of 10,000 lamports and no additional data. Rent is immediately debited from it on creation, resulting in a balance of 7,561 lamports:
```

> You can calculate like this:

Reviewer 3: ```suggestion
```

```

> 7,561 = 10,000 (= transfered lamports) - 2,439 (= this account's rent fee for a epoch)
> 2,439 = 19.055441478439427 (= rent rate) * 128 bytes (= minimum account size) * 1 (= epoch)

Reviewer 3: ```suggestion
Rent: 2,439 = 19.055441478439427 (rent rate) * 128 bytes (minimum account size) * 1 (epoch)
Account Balance: 7,561 = 10,000 (transfered lamports) - 2,439 (this account's rent fee for an epoch)
```

```

> And the account balance will be reduced to 5,122 lamports at the next epoch even if there is no activity:

Reviewer 3: ```suggestion
The account balance will be reduced to 5,122 lamports at the next epoch even if there is no activity:
```

```

> 5,122 = 7,561 (= current balance) - 2,439 (= this account's rent fee for a epoch)

Reviewer 3: ```suggestion
Account Balance: 5,122 = 7,561 (current balance) - 2,439 (this account's rent fee for an epoch)
```

```

> This also indicates an account will be immediately removed after creation if the transferred lamports is less than or equal to 2,439.

Reviewer 3: ```suggestion
Accordingly, a minimum-size account will be immediately removed after creation if the transferred lamports are less than or equal to 2,439.
```

## Rent exemption

> Alternatively, an account can be exempt from rent collection entirely by depositing a certain amount of lamports. Such a minimum amount is defined as the 2 years worth of rent fee. This is checked every time an account's balance is reduced and the rent is immediately debited once the balance goes below the minimum amount.

Reviewer 3: ```suggestion
Alternatively, an account can be made entirely exempt from rent collection by depositing at least 2 years-worth of rent. This is checked every time an account's balance is reduced and rent is immediately debited once the balance goes below the minimum amount.
```

> Program executable account required to be rent-exempt by the runtime to avoid being purged.

Reviewer 3: ```suggestion
Program executable accounts are required by the runtime to be rent-exempt to avoid being purged.
```

> Note: there is an RPC endpoint specifically to calculate this ([`getMinimumBalanceForRentExemption`](jsonrpc-api.md#getminimumbalanceforrentexemption)). Apps should rely on it. The following calculation is illustrative only.

Reviewer 3: ```suggestion
Note: Use the [`getMinimumBalanceForRentExemption` RPC endpoint](jsonrpc-api.md#getminimumbalanceforrentexemption) to calculate the minimum balance for a particular account size. The following calculation is illustrative only.
```

> For example, 105,290,880 lamports (=~ 0.105 SOL) is needed to be rent-exempt for a program executable with the size of 15,000 bytes:

Reviewer 3: ```suggestion
For example, a program executable with the size of 15,000 bytes requires a balance of 105,290,880 lamports (=~ 0.105 SOL) to be rent-exempt:
```

```

> 105,290,880 = 19.055441478439427 (= fee rate) * (128 + 15_000)(= account size including metadata) * ((365.25/2) * 2)(=epochs in 2 years)

Reviewer 3: ```suggestion
105,290,880 = 19.055441478439427 (fee rate) * (128 + 15_000)(account size including metadata) * ((365.25/2) * 2)(epochs in 2 years)
```

# Storage Rent for Accounts

Keeping accounts alive on Solana incurs a storage cost called _rent_ because the cluster must actively maintain the data to process any future transactions on it. This is different from Bitcoin and Ethereum, where storing accounts doesn't incur any costs.

The rent is debited from an account's balance by the runtime upon the first access (including the initial account creation) in the current epoch by transactions or once per an epoch if there are no transactions. The fee is currently a fixed rate, measured in bytes-times-epochs. The fee may change in the future.

> For the sake of simple rent calculation, minimum duration is a single epoch. The rent is not paid on a pro rata basis, meaning there is no fee/refund for a partial epoch rent. Thereby, the rent can be regarded to be paid upfront in full for each upcoming epoch. For newly-created accounts, this means the first rent isn't for the current partial epoch, but for the next full epoch, and subsequent rents are for further future epochs. If an already-rent-collected account balance remains to be below another rent fee, it'll be purged immediately at the start of the upcoming epoch.

Author: Here is a revamped explanation of the exact definition of rent duration. :)
